src/github.com/matrix-org/dendrite/clientapi/routing/register.go

```diff
@@ -212,16 +238,36 @@ func Register(
 			return util.MessageResponse(403, "HMAC incorrect")
 		}
 
-		return completeRegistration(req.Context(), accountDB, deviceDB, r.Username, r.Password, r.InitialDisplayName)
+		// Add SharedSecret to the list of completed registration stages
+		sessions[sessionID] = append(sessions[sessionID], authtypes.LoginTypeSharedSecret)
+
 	case authtypes.LoginTypeDummy:
 		// there is nothing to do
-		return completeRegistration(req.Context(), accountDB, deviceDB, r.Username, r.Password, r.InitialDisplayName)
+		// Add Dummy to the list of completed registration stages
+		sessions[sessionID] = append(sessions[sessionID], authtypes.LoginTypeDummy)
+
 	default:
 		return util.JSONResponse{
 			Code: 501,
 			JSON: jsonerror.Unknown("unknown/unimplemented auth type"),
 		}
 	}
+
+	// Check if a registration flow has been completed successfully
+	for _, flow := range cfg.Derived.Flows {
+		if checkFlowsEqual(flow, authtypes.Flow{sessions[sessionID]}) {
+			return completeRegistration(req.Context(), accountDB, deviceDB,
+				r.Username, r.Password, r.InitialDisplayName)
+		}
+	}
+
+	// There are still more stages to complete.
+	// Return the flows and those that have been completed.
+	return util.JSONResponse{
```

Comment:
@erikjohnston Did you mention that you typically use `return httputil.LogThenError(req, err)` for this rather than creating a `util.JSONResponse` on the spot?

Reply:
Only for when returning 500s. There may be other helper functions for doing 401

Reply:
But I think this is fine as is

Reply:
Cool, alright.